### ISSUE TITLE
Explicit projection results in incorrect plan

### ISSUE THREAD

**Issue by andygrove** [ballista-compute/ballista#320]
There is a bug in the DataFusion optimizer when explicit projections are addeded to a plan. The TPCH example reads too many columns from Parquet with the explicit projection added.

**Comment by jorgecarleitao** [comment]
I can look at it tomorrow - I am fairly familiar with the optimizers code base at this point.

**Comment by jorgecarleitao** [comment]
I have identified the root cause of this and I have a fix for this in DataFusion. Filed in [ARROW-9678](https://issues.apache.org/jira/browse/ARROW-9678)

**Comment by jorgecarleitao** [comment]
Actually, this is a separate issue.

The issue here is that when an expression in a projection is only used once in the rest of the plan, we can remove that column from the projection and just place the expression where it is used. When all expressions in a projection are used only once in the plan, we can just drop the projection altogether.

**Comment by jorgecarleitao** [comment]
I am surprised that this has a significant impact in performance, though. I.e. are we getting a significant performance difference between the following two hypothetical plans, @andygrove ?

```
Aggregate: groupBy=[[#a]], aggr=[[MAX(#b)]]
  Projection: complex(#b) as b, #a
    TableScan: test projection=Some([0, 1])";
```

```
Aggregate: groupBy=[[#a]], aggr=[[MAX(complex(#b))]]
    TableScan: test projection=Some([0, 1])";
```

There should not be such a difference with async: the aggregation does not need to wait for the projection to finish and will just receive record batches from the projection as it goes, right?

I would expect the cost of applying the identity projection to be `O(1)` on the number of records.

**Comment by andygrove** [comment]
The issue that caused me to file this was that the parquet reader was reading more columns than were needed by the query because of a bug in the projection push down logic.

No, I would not expect much difference between the two plans you posted. However, if there is a complex expression that is repeated (as in tpch query 1) then there could be benefit in having an explicit projection so that it just gets evaluated once.

**Comment by jorgecarleitao** [comment]
Got it, that is also my exact thinking.

My hypothesis that this is caused by projections being interpreted as being required by the plan, even when not all of its columns are required I.e. in a plan like this:

```
Aggregate: groupBy=[[#c]], aggr=[[MAX(#a)]]
  Selection: #c Gt Int32(1)
    Projection: #c, #a, #b
      TableScan: test projection=None
```

The optimizer currently scans the 3 columns, even though this can be optimized to

```
Aggregate: groupBy=[[#c]], aggr=[[MAX(#a)]]
  Selection: #c Gt Int32(1)
    Projection: #c, #a
      TableScan: test projection=Some([0, 2])";
```

i.e. the fact that a projection requires a column causes a scan, even when that column is ignored.

I PRed a fix to this here: https://github.com/apache/arrow/pull/7919
